File: seungeun/12주차/네트워크.py
def solution(n, computers):
    parent = list(range(n))
    
    def find(x):
        if parent[x] != x:
            parent[x] = find(parent[x])
        return parent[x]
    
    def union(a, b):
        ra = find(a)
        rb = find(b)
        
        if ra != rb:
            parent[rb] = ra
    
    for i in range(n):
        for j in range(n):
            if i != j and computers[i][j] ==1:
                union(i,j)
                
    return len(set(find(i) for i in range(n)))
    
    
    
    

File: yeongmok/12주차(유니온 파인)/네트워크.java
import java.util.*;

class Solution {
    int[] parent;
    public int solution(int n, int[][] computers) {
        parent = new int[n];

        for(int i=0;i<n;i++){
            parent[i] = i;
        }

        for(int i=0;i<computers.length;i++){
            for(int j=0;j<n;j++){
                if(computers[i][j]==1)
                    union(i,j);
            }
        }

        for(int i=0;i<parent.length;i++){
            parent[i] = find(i);
        }
        //System.out.println(Arrays.toString(parent));
        List<Integer> list = new ArrayList<>();
        for(int i=0;i<parent.length;i++){
            if(!list.contains(parent[i]))
                list.add(parent[i]);
        }

        return list.size();
    }

    int find(int x){
        if(parent[x]==x){
            return x;
        }
        return parent[x] = find(parent[x]);
    }

    void union(int i, int j){
        int rootA = find(i);
        int rootB = find(j);

        if(rootA!=rootB)
            parent[rootB] = rootA;

        return;
    }
}